### PR TITLE
feat(michelson): michelson optimisations

### PR DESCRIPTION
**Pull request by alanmarkoTrilitech**
# Context
https://app.asana.com/0/1207308081067689/1207308081144578

# Decription
This PR adds Michelson optimisations from SmartPy to LLTZ and thus achieves +-10% codesize compared to the original backend as tested on LIGO integration tests. It converts to and from Micheline <-> Michelson and utilises Michelson rewriter for modifications of the resulting code. Instead of using two different versions of Concat, one with 1 arg and another with 2 args, all concats are converted to list-based version.

# Manual testing
Build LIGO with this submodule and run tests

### CODE REVIEW

**Comment by rinderknecht** [comment]
When and how often do we have +10% increase in code size? That's not good.

**Comment by alanmarkoTrilitech** [comment]
@rinderknecht 

> When and how often do we have +10% increase in code size? That's not good.

I don't see how is it "not good". This compilation is different than what the original LIGO backend did, compiling from a different IR, so it would be quite surprising if we matched the codesize precisely or improve it in all cases (I would actually argue it isn't necessary). Compared to LLTZ without optimisations it is often 50%+ improvement (estimate, haven't measured exactly). I haven't looked exactly yet into the cases that cause the increase in code size, but I would imagine it is things like e.g. the Concat - where I first make a list to to unify args and the apply instruction, instead of using same instruction with two different arg counts.

**Comment by johnyob** [comment]
Why do we need so many files? Surely we can take a subset of these. Additionally (aside from Tezos-specific stuff), we should be relying on `Core` as a std library. This PR introduces changes that are against that.

**Comment by johnyob** [comment]
Additionally before porting optimisations, we should have infrastructure for measuring the improvements (automatically as tests)

**Comment by alanmarkoTrilitech** [comment]
> Additionally before porting optimisations, we should have infrastructure for measuring the improvements (automatically as tests)

> I wrote a code that measures it via Of_michelson.measure, which isn't part of this PR. There of course will be tests measuring code size, though I think they can be included after porting these optimisations. Currently we also don't measure the code size of lltz in tests, but I can see by testing the contracts roughly what is the difference between LLTZ and original backend.

**Comment by alanmarkoTrilitech** [comment]
well it is mainly because of the utils, I guess you are right I can choose a smaller subset of. Except for the smartPy michelson optimisations, the code still relies on the Core as std library.

**Comment by rinderknecht** [comment]
Users pay for the generated code size, and the user base will be that of SmarPy + LIGO. Ten percent is a sensible threshold, and if +10% in code size is rare, then it's fine. I guess that a decision has to be made whether we care first about users or devs, and what are the acceptable trade-offs.

**Comment by alanmarkoTrilitech** [comment]
> Users pay for the generated code size, and the user base will be that of SmarPy + LIGO. Ten percent is a sensible threshold, and if +10% in code size is rare, then it's fine. I guess that a decision has to be made whether we care first about users or devs, and what are the acceptable trade-offs.

well, we can make these decisions regarding whether it is acceptable to have sometimes 10% increase in the final release (also I want to note that it sometimes it results in 10% decrease, so it isn't always. worse),  however for a prototype that is behind --lltz-ir flag I think it is fine and also was obvious from the start that it won't be precisely matching the code sizes right away.

If we want to have it always better, it anyway needs to first get to a state like this where it is close enough, before it gets further optimised until it is always better.

**Comment by rinderknecht** [comment]
I agree.

**Comment by johnyob** [comment]
> I wrote a code that measures it via Of_michelson.measure, which isn't part of this PR. There of course will be tests measuring code size, though I think they can be included after porting these optimisations. 

I'd disagree, the code for measuring code size and instrumenting our tests with it would be beneficial _before_ including these optimisations. Since we get a base measurement + then we can actually verify in which cases the optimisations improve codesize and which ones don't

**Comment by alanmarkoTrilitech** [comment]
you can still get pre and post optimisations measurement even after inclusion of optimisations as it is behind lltz-ir flag. for example the code I have for doing it computes both at once while running tests and outputs it to a file.

File: lib/michelson/optimisations/smartpy_lltz_interopt.ml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+module Tezos_micheline = Tezos_micheline
+module Michelson = Michelson
+
+let rec micheline_to_oasis_micheline (node : (unit, Michelson.Ast.Prim.t) Tezos_micheline.Micheline.node) : Oasis_core.Micheline.t =
+  match node with
+  | Int (_, z) ->
```
Comment (alanmarkoTrilitech) [inline lib/michelson/optimisations/smartpy_lltz_interopt.ml:6]: add unit

**Comment by rinderknecht** [comment]
What's the difference between Oasis Micheline and Micheline?

**Comment by alanmarkoTrilitech** [comment]
> What's the difference between Oasis Micheline and Micheline?
There is no difference between them, but SmartPy used oasis micheline instead of micheline for legacy reasons. Possibly these could be substituted directly, though the same is achieved with just having this conversion.

**Review by rinderknecht — APPROVED** [review]

**Review by rinderknecht — APPROVED** [review]

**Pull request merged**